tsl/src/planner.c

```diff
@@ -146,3 +149,82 @@ tsl_set_rel_pathlist(PlannerInfo *root, RelOptInfo *rel, Index rti, RangeTblEntr
 
 	ts_cache_release(hcache);
 }
+
+/*
+ * Check session variable to disable distributed insert with COPY.
+ *
+ * Allows testing and comparing different insert plans. Not a full GUC, since
+ * in most cases this is not something that users are expected to disable.
+ */
+static bool
+copy_mode_enabled(void)
+{
+	const char *enable_copy =
+		GetConfigOption("timescaledb.enable_distributed_insert_with_copy", true, false);
+
+	/* Default to enabled */
+	if (NULL == enable_copy)
+		return true;
+
+	return strcmp(enable_copy, "true") == 0;
+}
+
+/*
+ * Decide on a plan to use for distributed inserts.
+ */
+Path *
+tsl_create_distributed_insert_path(PlannerInfo *root, ModifyTablePath *mtpath, Index hypertable_rti,
+								   int subplan_index)
+{
+	bool copy_possible = copy_mode_enabled();
+
+	/* Check if it is possible to use COPY in the backend.
+	 *
+	 * There are two cases where we cannot use COPY:
+	 *
+	 * 1. ON CONFLICT clause exists
+	 *
+	 * 2. RETURNING clause exists and tuples are expected to be modified on
+	 *    INSERT by a trigger.
+	 *
+	 * For case (2), we assume that we can return the original tuples if there
+	 * are no triggers on the root hypertable (we also assume that chunks on
+	 * data nodes aren't modified with their own triggers). Conversely, if
+	 * there are any non-known triggers, we cannot use COPY in the backend
+	 * when there is a RETURNING clause.
+	 */
+	if (copy_possible)
+	{
+		if (NULL != mtpath->onconflict)
+			copy_possible = false;
+		else if (NIL != mtpath->returningLists)
+		{
+			const RangeTblEntry *rte = planner_rt_fetch(hypertable_rti, root);
+			const Relation rel = table_open(rte->relid, AccessShareLock);
+			int i;
+
+			for (i = 0; i < rel->trigdesc->numtriggers; i++)
+			{
+				const Trigger *trig = &rel->trigdesc->triggers[i];
+
+				/*
+				 * Check for BEFORE INSERT triggers as those are the ones that can
+				 * modify a tuple.
+				 */
+				if (strcmp(trig->tgname, INSERT_BLOCKER_NAME) != 0 &&
```

Comment:
So trigger checking happens on the hypertable on the AN, because foreign tables will not have any triggers on them?

Reply:
Yes, and we assume that the chunks on DNs have the same triggers. This is the best we can do. This might warrant having a GUC to disable this plan node, but not sure it is necessary at this point (currently providing an anonymous session variable).